text/0813-template-layering.md

```diff
@@ -0,0 +1,984 @@
+depend on a suitable polyfill.
+
+## Detailed design
+
+This RFC proposes to re-specify the `<template>` language extension into a
+de-sugaring into `template()` calls:
+
+1. Top-level declaration
+
+   ```gjs
+   import Foo from "somewhere";
+
+   <template>
+     <Foo />
+   <template>
+   ```
+
+   becomes
+
+   ```js
+   import { template } from "@ember/template-compilation";
+   import Foo from "somewhere";
+
+   export default template(
+     `
+     <Foo />
+   `,
+     () => ({ Foo })
+   );
+   ```
+
+2. Expression
+
+   ```gjs
+   import Foo from "somewhere";
+
+   export const Bar = <template><Foo /><template>;
+   ```
+
+   becomes
+
+   ```js
+   import { template } from "@ember/template-compilation";
+   import Foo from "somewhere";
+
+   export const Bar = template(`<Foo />`, () => ({ Foo }));
+   ```
+
+3. Class
+
+   ```gjs
+   import Foo from "somewhere";
+
+   export default class Bar {
+     <template><Foo /><template>
+   }
+   ```
+
+   becomes
+
+   ```js
+   import { template } from "@ember/template-compilation";
+   import Foo from "somewhere";
+
+   export default class Bar {
+     static {
+       template(`<Foo />`, () => ({ Foo }), this);
+     }
+   }
+   ```
+
+   When used in this form, the function returns/passes through the target
+   supplied in the third argument (the component class).
+
+   Note that the use of a static initializer block here is important for
+   capturing access to private fields. This will be discussed in a subsequent
+   section.
+
+In these small snippets, the scope bindings may look very verbose compared to
+the size of the templates, but in real-world templates, the ratio will improve.
+
+This de-sugaring specified in this section is intended to **subsume** the parts
+of RFC #779 which references its compilation output, and where it specified the
+feature in terms of the low-level primitives such as `precompileTemplate()`. If
+this RFC is accepted, the `<template>` feature should only be understood in
+terms of `template()`, which will be a stable and public API that stands on its
+own.
+
+Any build-time optimization that pre-compiles the `template()` calls at build
+time are required to ensure their output to have equivalent semantics as the
+runtime `template()` calls, assuming the same set of AST plugins are applied.
+Their exact compilation output should be considered private implementation
+details which may change over time and should not be relied upon.
+
+Once this RFC is accepted, any future changes (including deprecations) to the
+primitive APIs referenced in RFC #779 should have no direct bearing on the
+`<template>` or `template()` features.
+
+### Tagged Template Literals
+
+Alternatively, the `template()` function can also be used as a tag function for
+a JavaScript tagged template string. This alternate form automatically captures
+the lexical scope variable bindings. However, this form does not have a runtime
+implementation and must be processed by a build step.
```

Comment:
This *does not* actually "work" in any of the target environments described above. For example, in CodePen, the code would parse, but you wouldn't be able to *use* it at all: The template part of it would be completely inert! I’m curious what you see this providing for users?